src-java/kilda-model/src/main/java/org/openkilda/model/HaFlow.java

```diff
@@ -255,6 +257,46 @@ private void validateReversePath(HaFlowPath haPath) {
         }
     }
 
+    /**
+     * Get up to 3 endpoint switchIds.
+     */
+    public Set<SwitchId> getEndpointSwitchIds() {
+        Set<SwitchId> switchIds = Sets.newHashSet(getSharedSwitchId());
+        for (HaSubFlow subFlow : getSubFlows()) {
+            switchIds.add(subFlow.getEndpointSwitchId());
+        }
+        return switchIds;
+    }
+
+    /**
+     * Return opposite pathId to passed pathId.
+     */
+    public Optional<PathId> getOppositePathId(@NonNull PathId pathId) {
+        if (pathId.equals(getForwardPathId()) && getReversePathId() != null) {
+            return Optional.of(getReversePathId());
+        } else if (pathId.equals(getReversePathId()) && getForwardPathId() != null) {
+            return Optional.of(getForwardPathId());
+        } else if (pathId.equals(getProtectedForwardPathId()) && getProtectedReversePathId() != null) {
+            return Optional.of(getProtectedReversePathId());
+        } else if (pathId.equals(getProtectedReversePathId()) && getProtectedForwardPathId() != null) {
+            return Optional.of(getProtectedForwardPathId());
+        } else {
+            // Handling the case of non-active paths.
+            Optional<Long> requestedPathCookie = getPath(pathId)
+                    .map(HaFlowPath::getCookie)
+                    .map(FlowSegmentCookie::getFlowEffectiveId);
+            if (requestedPathCookie.isPresent()) {
+                return getPaths().stream()
+                        .filter(path -> !path.getHaPathId().equals(pathId))
```

Comment:
shouldn't you filter out the protected path as well?

Reply:
No. this method must return opposite protected path too. 
Main and protected path have different cookies.

Reply:
alright, shouldn't you then include that different cookie? 🙂
I mean in the section above it returns the opposite of main or the opposite of protected, but else handles only the main. The method doesn't need to return the opposite of a protected non-active path?

Reply:
Why do you think that `else` handles only the main?
Else section tries to find a path by cookie. main and protected paths have different cookies. So else section handles both of them